File: selenium/src/java/org/openqa/selenium/internal/selenesedriver/FindElement.java
/*
Copyright 2007-2009 WebDriver committers
Copyright 2007-2009 Google Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package org.openqa.selenium.internal.selenesedriver;

import com.google.common.base.Charsets;
import com.google.common.base.Throwables;
import com.google.common.collect.ImmutableMap;
import com.thoughtworks.selenium.Selenium;
import com.thoughtworks.selenium.SeleniumException;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebDriverException;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.Map;

public class FindElement implements SeleneseFunction<Map<String, String>> {

  private final static String SCRIPT =
      "var e = selenium.browserbot.findElement('%s');" +
      "bot.inject.cache.addElement(e);";

  private long implicitlyWait = 0;


  public Map<String, String> apply(Selenium selenium, Map<String, ?> args) {
    String how = (String) args.get("using");
    String using = (String) args.get("value");

    String locator = null;
    if ("class name".equals(how)) {
      locator = "css=." + using;
    } else if ("css selector".equals(how)) {
      locator = "css=" + using;
    } else if ("id".equals(how)) {
      locator = "id=" + using;
    } else if ("link text".equals(how)) {
      locator = "link=" + using;
    } else if ("name".equals(how)) {
      locator = "name=" + using;
    } else if ("tag name".equals(how)) {
      locator = "xpath=//" + using;
    } else if ("xpath".equals(how)) {
      locator = "xpath=" + using;
    } else {
      throw new WebDriverException("Cannot determine locator mechanism from: " + how);
    }

    if (locator != null) {
      locator = String.format(SCRIPT, locator.replaceAll("'", "\\\\'"));

      long startTime = System.currentTimeMillis();
      do {
        try {
          String key = selenium.getEval(locator);

          key = URLEncoder.encode(key, "utf-8");

          return ImmutableMap.of("ELEMENT", "stored=" + key);
        } catch (SeleniumException e) {
          // Ignore. The element couldn't be found
        } catch (UnsupportedEncodingException e) {
          // This really can't happen on a conforming JVM
          throw Throwables.propagate(e);
        }
      } while (System.currentTimeMillis() - startTime <= implicitlyWait);
    }

    throw new NoSuchElementException(
        String.format("Cannot find element using %s=%s ", how, using));
  }

  public ImplicitWait implicitlyWait() {
    return new ImplicitWait();
  }

  public class ImplicitWait implements SeleneseFunction<Object> {
    public Object apply(Selenium selenium, Map<String, ?> args) {
      FindElement.this.implicitlyWait = ((Number) args.get("ms")).longValue();
      return null;
    }
  }
}

File: javascript/webdriver-atoms/inject/find_element.js
// Copyright 2011 WebDriver committers
// Copyright 2011 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * @fileOverview Ready to inject atoms into the page.
 */

goog.provide('webdriver.inject');

goog.require('bot.inject');
goog.require('bot.locators');


/**
 * Finds an element by using the given lookup strategy.
 * @strategy {string} stategy The strategy to use to locate the element.
 * @param {string} using The locator to use.
 * @param {(Document|Element)=} opt_root The document or element to perform
 *     the search under. If not specified, will use {@code document}
 *     as the root.
 * @return {string} The script result wrapped
 *     in a JSON string as defined by the WebDriver wire protocol.
 */
webdriver.inject.findElement = function(strategy, using, opt_root) {
  var locator = {}
  locator[strategy] = using;
  return bot.inject.executeScript(bot.locators.findElement,
                           [locator, opt_root], true);
}
